### PR TITLE
Fix S4275: FP when the value is encapsulated in another class

### LINKED ISSUE #2867: Fix S4275: FP when the value is encapsulated in another class
### Description

When the field represents a class which encapsulates the actual value returned from the property, S4275 is incorrectly raised.

### Repro steps

Consider this code:
```
public class Sample
{
    private class ValueWrapper<T>
    {
        public T Value { get; set; }
        // ...
    }

    private readonly ValueWrapper<double> _someMember = new ValueWrapper<double>();

    public double SomeMember
    {
        get => _someMember.Value;
        set => _someMember.Value = value;
    }
}
```

### Expected behavior

S4275 should not be raised because the setter _does_ access the corresponding field.

### Actual behavior

S4275 is raised on the setter.

### Related information

* SonarLint Version 4.14.0.12332
* Visual Studio Version 16.4.0

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #2867

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs [hunk 1]
```diff
@@ -54,7 +54,9 @@ protected override IEnumerable<FieldData> FindFieldAssignments(IPropertySymbol p
                 FieldData? foundField = null;
                 if (node is AssignmentExpressionSyntax assignment && node.IsKind(SyntaxKind.SimpleAssignmentExpression))
                 {
-                    foundField = ExtractFieldFromExpression(AccessorKind.Setter, assignment.Left, compilation);
+                    foundField = assignment.Left.DescendantNodesAndSelf().OfType<ExpressionSyntax>()
+                        .Select(x => ExtractFieldFromExpression(AccessorKind.Setter, x, compilation))
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs]: pedantic: "x" does not provide any helpful info, I think that using "expression" would provide a bit of context

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/PropertiesAccessCorrectField.cs [hunk 2]
```diff
@@ -57,7 +57,9 @@ protected override IEnumerable<FieldData> FindFieldAssignments(IPropertySymbol p
                 FieldData? foundField = null;
                 if (node is AssignmentStatementSyntax assignment && assignment.IsKind(SyntaxKind.SimpleAssignmentStatement))
                 {
-                    foundField = ExtractFieldFromExpression(AccessorKind.Setter, assignment.Left, compilation);
+                    foundField = assignment.Left.DescendantNodesAndSelf().OfType<ExpressionSyntax>()
```
Comment (costin-zaharia-sonarsource) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/PropertiesAccessCorrectField.cs]: The code is similar with the one for C#. Can we extract a few generic methods in order to enable reuse?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs]: I don't want to change this. It's lambda, so single letter variable is fine. `x` is by convention the most used one (I've found 63 other files in project with `x =>`). 
It comes from `.OfType<ExpressionSyntax>()` so in this case, it's very clear and explicit what `x` is.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/PropertiesAccessCorrectField.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/PropertiesAccessCorrectField.cs]: I didn't like the old version of the rule and considered full refactoring. But there are too many language dependent types => too much work for this sprint. So I went for keeping the original scheme of the rule.
Turning VB into low hanging fruit is a bigger topic.

**Pull request merged**